service/src/event/ordering_task.rs

```diff
@@ -365,112 +434,96 @@ impl OrderingState {
-        }
-        let found = undelivered.len();
-        let mut new = 0;
-        for (key, data) in undelivered {
-            let event_cid = key.cid().ok_or_else(|| {
-                Error::new_invalid_arg(anyhow::anyhow!("EventID is missing a CID: {}", key))
+        event_data: Vec<(EventId, Vec<u8>)>,
+    ) -> Result<usize> {
+        trace!(cnt=%event_data.len(), "Processing undelivered events batch");
+        let mut to_store_asap = Vec::new();
+        let mut event_cnt = 0;
+        for (event_id, carfile) in event_data {
+            let event_cid = event_id.cid().ok_or_else(|| {
+                Error::new_invalid_arg(anyhow::anyhow!("EventID is missing a CID: {}", event_id))
             })?;
-            let (insertable_body, maybe_prev) =
-                CeramicEventService::parse_event_carfile(event_cid, &data).await?;
-            if let Some(prev) = maybe_prev {
-                let event = DeliverableEvent::new(insertable_body.cid, prev, None);
-                if self.track_pending(event).is_some() {
-                    new += 1;
+
+            let loaded = CeramicEventService::parse_event_carfile_cid(event_cid, &carfile).await?;
+
+            let event = match &loaded.header {
+                EventHeader::Init { cid, .. } => {
+                    warn!(%cid,"Found init event in database that wasn't previously marked as deliverable. Updating now...");
+                    to_store_asap.push(*cid);
+                    StreamEvent::InitEvent(*cid)
                 }
-            } else {
-                // safe to ignore in tests, shows up because when we mark init events as undelivered even though they don't have a prev
-                info!(event_cid=%insertable_body.cid, "Found undelivered event with no prev while processing undelivered. Should not happen. Likely means events were dropped before.");
-                self.ready_events.push_back(insertable_body.cid);
-                new += 1; // we treat this as new since it might unlock something else but it's not actually going in our queue is it's a bit odd
-            }
+                EventHeader::Data { cid, prev, .. } | EventHeader::Time { cid, prev, .. } => {
+                    StreamEvent::Undelivered(StreamEventMetadata::new(*cid, *prev))
+                }
+            };
+
+            event_cnt += 1;
+            self.add_stream_event(loaded.header.stream_cid(), event);
         }
-        trace!(%new, %found, "Adding undelivered events to pending set");
-        Ok((new, found))
-    }
 
-    fn add_incoming_batch(&mut self, events: Vec<DeliverableEvent>) -> HashSet<InitCid> {
-        let mut updated_streams = HashSet::with_capacity(events.len());
-        for event in events {
-            if let Some(updated_stream) = self.track_pending(event) {
-                updated_streams.insert(updated_stream);
+        if !to_store_asap.is_empty() {
+            info!("storing init events that were somehow missed previously");
```

Comment:
again, this is part of why I feel like these programming bugs should be fatal, so we don't then have this random code sprinkled around to handle a case we think should never actually happen.